stl/inc/xutility

```diff
@@ -3714,6 +3765,56 @@ namespace ranges {
         }
     };
 
+    // clang-format off
+    template <forward_iterator _Wrapped, class _Se>
+        requires sentinel_for<remove_cvref_t<_Se>, _Wrapped>
+    _NODISCARD constexpr auto _Get_final_iterator_unwrapped(const _Unwrapped_t<_Wrapped>& _UFirst, _Se&& _Last) {
+        // clang-format on
+        // find the iterator in [_UFirst, _Get_unwrapped(_Last)) which equals _Get_unwrapped(_Last) [possibly O(N)]
+        auto _ULast = _Get_unwrapped(_STD forward<_Se>(_Last));
+        if constexpr (is_same_v<remove_cvref_t<_Se>, _Wrapped>) {
+            return _ULast;
+        } else if constexpr (sized_sentinel_for<remove_cvref_t<_Se>, _Wrapped>) {
+            return _RANGES next(_UFirst, _ULast - _UFirst);
+        } else {
+            return _RANGES next(_UFirst, _STD move(_ULast));
+        }
+    }
+
+    template <forward_range _Rng>
+    _NODISCARD constexpr auto _Get_final_iterator_unwrapped(_Rng& _Range) {
+        // find the (unwrapped) iterator in _Range which equals _Uend(_Range) [possibly O(N)]
+        if constexpr (common_range<_Rng>) {
+            if constexpr (same_as<_Unwrapped_t<_RANGES iterator_t<_Rng>>, decltype(_Uend(_Range))>) {
+                return _Uend(_Range);
+            } else {
+                return _Get_unwrapped(_RANGES end(_Range));
+            }

```

Comment:
> I don't understand why these would be distinct; when does a type's `._Unwrapped_end()` return a different type to `_Get_unwrapped(_RANGES end(_Range))`? Shouldn't that be library UB?

This was the case until we recently threw `forward_list` at this for the first time. `forward_list::_Unchecked_end` returns a sentinel type because Billy (I was _so_ proud) found it resulted in better codegen than `_Get_unwrapped(forward_list::end())`, which the optimizer doesn't realize is always a null pointer. I had completely forgotten this when extending the iterator unwrapping machinery for Ranges.

It's not a technique I'd suggest for new code, but it's just handy enough for janky old code to make it work.
 
> I'll also point out the other overload of `_Get_final_iterator_unwrapped` (the one which takes a midpoint iterator) does not do this weirdness.

That's completely bogus. I'll fix.
 
> edit: ah, this is pre-existing, and is just being moved from `<algorithm>`, so no change requested. It should be fixed after we merge this tho.

Meh, it's four lines - barely even noise in this PR.